providers/unleash/src/main/java/dev/openfeature/contrib/providers/unleash/ContextTransformer.java

```diff
@@ -0,0 +1,75 @@
+package dev.openfeature.contrib.providers.unleash;
+
+import dev.openfeature.sdk.EvaluationContext;
+import dev.openfeature.sdk.ImmutableContext;
+import dev.openfeature.sdk.Value;
+import io.getunleash.UnleashContext;
+
+import java.time.ZonedDateTime;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Transformer from Unleash context to OpenFeature context and vice versa.
+ */
+public class ContextTransformer {
+
+    public static final String CONTEXT_APP_NAME = "appName";
+    public static final String CONTEXT_USER_ID = "userId";
+    public static final String CONTEXT_ENVIRONMENT = "environment";
+    public static final String CONTEXT_REMOTE_ADDRESS = "remoteAddress";
+    public static final String CONTEXT_SESSION_ID = "sessionId";
+    public static final String CONTEXT_CURRENT_TIME = "currentTime";
+
+    protected static UnleashContext transform(EvaluationContext ctx) {
+        UnleashContext.Builder unleashContextBuilder = new UnleashContext.Builder();
+        ctx.asObjectMap().forEach((k, v) -> {
+            switch (k) {
+                case CONTEXT_APP_NAME:
+                    unleashContextBuilder.appName(String.valueOf(v));
+                    break;
+                case CONTEXT_USER_ID:
+                    unleashContextBuilder.userId(String.valueOf(v));
+                    break;
+                case CONTEXT_ENVIRONMENT:
+                    unleashContextBuilder.environment(String.valueOf(v));
+                    break;
+                case CONTEXT_REMOTE_ADDRESS:
+                    unleashContextBuilder.remoteAddress(String.valueOf(v));
+                    break;
+                case CONTEXT_SESSION_ID:
+                    unleashContextBuilder.sessionId(String.valueOf(v));
+                    break;
+                case CONTEXT_CURRENT_TIME:
+                    unleashContextBuilder.currentTime(ZonedDateTime.parse(String.valueOf(v)));
+                    break;
+                default:
+                    unleashContextBuilder.addProperty(k, String.valueOf(v));
+                    break;
+            }
+        });
+        return unleashContextBuilder.build();
+    }
+
+    /**
+     * Transform UnleashContext to EvaluationContext.
+     * @param unleashContext the UnleashContext
+     * @return transformed EvaluationContext
+     */
+    public static EvaluationContext transform(UnleashContext unleashContext) {
```

Comment:
I think this is only needed for the tests?

Reply:
Not only for tests.
By [Usage example](https://github.com/open-feature/java-sdk-contrib/blob/8a6316705d8ac7779ad2f00db4b18710120de7d0/providers/unleash/README.md#usage-example):
```
UnleashContext unleashContext = UnleashContext.builder().userId("1").build();
EvaluationContext evaluationContext = EvaluationContext.transform(unleashContext);
featureEnabled = client.getBooleanValue(FLAG_NAME, false, evaluationContext);
```

What do you think ?

Reply:
why would you use the Open Feture wrapper if you use the Unleash SDK? 

Imh the Open Feature API should use generics for context containers, but they don't.

Reply:
Oh that's interesting, I think I have the same worries as ivarconr - that's not really about this implementation but about the way OpenFeature is intended to be used. If I have to put logic specific to my vendor SDK in the code, that sounds like it would be tough to migrate to another provider, since I'd have to find all these cases and map them to a new provider's content implementation. Totally might have missed something obvious here, since I don't know OpenFeature very well (yet!)

Reply:
This is a valid point. How I see it, evaluation context, like _USER_ID_ is something for a specific provider service, which not always used or defined by this key. Then context can be for attributes specific for this provider. Since such context properties are not defined at _OpenFeature_, then I think this part is sort of provider specific. As for using _UnleashContext_, it is not mandatory, user can set attributes map keys manually, but I think as it sort of custom anyway, it can be simpler for the user.

CC @toddbaert

Reply:
> why would you use the Open Feture wrapper if you use the Unleash SDK?
> 
> Imh the Open Feature API should use generics for context containers, but they don't.

> Oh that's interesting, I think I have the same worries as ivarconr - that's not really about this implementation but about the way OpenFeature is intended to be used. If I have to put logic specific to my vendor SDK in the code, that sounds like it would be tough to migrate to another provider, since I'd have to find all these cases and map them to a new provider's content implementation. Totally might have missed something obvious here, since I don't know OpenFeature very well (yet!)

Sorry I missed this until now!

@sighphyre and @ivarconr are correct about this. It's an anti-pattern to expect users to interact with an `UnleashContext` directly if they're using open feature.

If the `userId` is a special field that consistently identifies a user in Unleash, it should be mapped from the [targetingKey](https://openfeature.dev/specification/sections/evaluation-context#requirement-311). This is the only rigorously defined entry in the `EvaluationContext`. The transformation should be done internally. If you want to give users some control over it, you could expose a transformer function in your provider constructor to allow customization of this behavior.

Reply:
hi @toddbaert,
Unleash context contains several optional parameters, user ID is one of them.
Then instead of:
```
UnleashContext unleashContext = UnleashContext.builder()
            .userId(userIdValue)
            .currentTime(currentTimeValue)
            .sessionId(sessionIdValue)
            .remoteAddress(remoteAddressValue)
            .environment(environmentValue)
            .appName(appNameValue)
            .addProperty(customPropertyKey, customPropertyValue)
            .build();
EvaluationContext evaluationContext = ContextTransformer.transform(unleashContext);
```
should I replace the usage by exposing context keys constants, with something like:
```
Map<String, Value> attributes = new HashMap<>();
        attributes.put(CONTEXT_APP_NAME, new Value((appName));
        attributes.put(CONTEXT_USER_ID, new Value((userId));
        attributes.put(CONTEXT_ENVIRONMENT, new Value((env));
        attributes.put(CONTEXT_SESSION_ID, new Value((sessionId));
        attributes.put(CONTEXT_REMOTE_ADDRESS, new Value((remoteAddress));
        attributes.put(CONTEXT_CURRENT_TIME, new Value((currentTime));
        return new ImmutableContext(attributes);
```
This looks like a better usage, this is what you mean ?
Parameters are optional, so possibly only some are enough for most cases.

I can also create _UnleashProviderContext_, it may be over-implementation.

As for whether user ID should be mapped to targeting key - can be either parameter like others or targeting key. @ivarconr ?

Reply:
_if_ there is a dedicated field in Unleash that denotes the user of the flag evaluation, that should be what's mapped from `targetingKey`, especially if it's used to maintain stickiness of random assignments, etc. If there's no such value, then there's no need for this mapping.

For the other values, I would recommend you _don't_ use exported constants, because that again creates a dependency in all flag-related application code directly on your provider. Instead, I would perhaps recommend a mapping function (a [Consumer](https://docs.oracle.com/javase/8/docs/api/java/util/function/Consumer.html) perhaps) that can be supplied when you construct the provider. It would take an `EvaluationContext` and return a `UnleashedContext`. Then consumers could map the values exactly as they want by supplying this param. You could have a default that just maps them using the same keys (`currentTimeValue`, `sessionIdValue`, etc).

Reply:
> As for whether user ID should be mapped to targeting key - can be either parameter like others or targeting key. I think it is not mandatory, then regular context parameter is better fit than targeting key here, but I am not sure. @ivarconr ?

Correct. All fields in the unleash context are optional, but a few might be set for you if you do not set them yourself (environment and appName are examples). In fact a user can fully remove "userId" as a possible target in the Unleash API.

Reply:
I updated the code and usage according to the comments.